server/vertx/src/main/scala/sttp/tapir/server/vertx/decoders/VertxInputDecoders.scala

```diff
@@ -0,0 +1,127 @@
+package sttp.tapir.server.vertx.decoders
+
+import java.io.{ByteArrayInputStream, File}
+import java.nio.ByteBuffer
+import java.util.Date
+
+import io.vertx.lang.scala.VertxExecutionContext
+import io.vertx.scala.ext.web.RoutingContext
+import sttp.model.Part
+import sttp.tapir.internal.Params
+import sttp.tapir.server.internal.{DecodeInputs, DecodeInputsResult, InputValues, InputValuesResult}
+import sttp.tapir.server.vertx.VertxEndpointOptions
+import sttp.tapir.server.vertx.encoders.VertxOutputEncoders
+import sttp.tapir.server.vertx.handlers.tryEncodeError
+import sttp.tapir.server.{DecodeFailureContext, DecodeFailureHandling}
+import sttp.tapir.{DecodeResult, Endpoint, EndpointIO, RawBodyType}
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.reflect.ClassTag
+import scala.util.Random
+
+object VertxInputDecoders {
+
+  /**
+   * Decodes the inputs, the body if needed, and if it succeeds invokes the logicHandler
+   * @param endpoint the endpoint definition
+   * @param rc the RoutingContext
+   * @param logicHandler callback to execute if the inputs are decoded properly
+   * @param endpointOptions endpoint options (execution context, etc.)
+   * @param ect the error class to convert to, if the user recovers error
+   * @tparam E Error parameter type, if it should be caught and handled by the user
+   */
+  private [vertx] def decodeBodyAndInputsThen[E](
+    endpoint: Endpoint[_, E, _, _],
+    rc: RoutingContext,
+    logicHandler: Params => Unit,
+  )(implicit endpointOptions: VertxEndpointOptions, ect: Option[ClassTag[E]]): Unit = {
+    decodeBodyAndInputs(endpoint, rc).map {
+      case values: DecodeInputsResult.Values =>
+        InputValues(endpoint.input, values) match {
+          case InputValuesResult.Value(params, _) =>
+            logicHandler(params)
+          case InputValuesResult.Failure(_, failure) =>
+            tryEncodeError(endpoint, rc, failure)
+        }
+      case DecodeInputsResult.Failure(input, failure) =>
+        val decodeFailureCtx = DecodeFailureContext(input, failure)
+        endpointOptions.decodeFailureHandler(decodeFailureCtx) match {
+          case DecodeFailureHandling.NoMatch =>
+            endpointOptions.logRequestHandling.decodeFailureNotHandled(endpoint, decodeFailureCtx)(endpointOptions.logger)
+            rc.response.setStatusCode(404).end()
+          case DecodeFailureHandling.RespondWithResponse(output, value) =>
+            endpointOptions.logRequestHandling.decodeFailureHandled(endpoint, decodeFailureCtx, value)(endpointOptions.logger)
+            VertxOutputEncoders.apply(output, value)(endpointOptions)(rc)
+        }
+    }(endpointOptions.executionContextOr(VertxExecutionContext(rc.vertx.getOrCreateContext)))
```

Comment:
This seems to occur in a couple of places - maybe this EC-providing method could be extracted somewhere?>